### DISCUSSION TITLE
Feature request: Export/import settings to Obsidian

### DISCUSSION THREAD

**Discussion by Mikle-Bond** [coddingtonbear/obsidian-web#114] [Ideas]
I examined the Local Rest API spec, and noticed, that with `application/vnd.olrapi.note+json` content type, you can receive the frontmatter as JSON. So we could store settings, as they are being exported from extension, inside one of the notes in the frontmatter, as YAML.

I did a little test. I exported settings from Obsidian Web, into `obsidian-local-rest-api.settings.json`, and then run
```shell
cat obsidian-local-rest-api.settings.json | yq -y '{ "obsidian-web" : . }' | clip
```
to get YAML representation. I pasted that into a new note (`util/obsidian-web-settings.md`) and verified it like this
```shell
curl -k -H 'accept: application/vnd.olrapi.note+json' -H "Authorization: Bearer $token" https://localhost:27124/vault/util/obsidian-web-settings.md | jq '.frontmatter."obsidian-web"' > out.json
diff -w obsidian-local-rest-api.settings.json out.json
```
(With `-w` key, because exported file has CRLF endings, and `jq` emits LF endings)

We could instruct the browser extension to search for a note with specified frontmatter key (similar to how it searches for `url` keys, and ask user whether it should use that as a storage for its settings. So that new browsers could pick it up seamlessly.

**Comment by coddingtonbear** [comment]
This is a clever idea, and I'm up for at least starting a discussion about what a PR adding this functionality would have to include if you're interested in developing such a solution.

**Comment by 0xdade** [comment]
Bringing my idea from https://github.com/coddingtonbear/obsidian-web/discussions/177 over here, another option could be to just drop the json file into a path in a vault, using a code block to store the json. This could be done fairly easily with the existing local REST API integration, I think, and we wouldn't have to account for converting to yaml and back.

I may look into this in the coming weeks -- I'm a big fan of Obsidian Web and would love to make sure my settings are consistently available across Chrome profiles and across my devices.

**Reply by coddingtonbear** [reply]
If you do, I'd recommend maybe talking over a rough plan of how you are planning to accomplish it with me first.  Time, for me, is very hard to come by, and as a result I have to be very selective about what kinds of complexity I'm up for maintaining.  I wouldn't want you to spend your time making a feature I wouldn't be comfortable with merging.

**Reply by Mikle-Bond** [reply]
The codeblock as the storage is a good idea. 
Back when I posted the request, Obsidian didn't have the Properties plugin yet. It seemd natural to store stuff in the frontmatter, because it is well-structured part of a markdown note, and back then Obsidian wouldn't mess it up. This day (Obsidian v1.5.8), I am not so sure. Seems like putting nested objects into front matter makes Properties view angry. And if you click on anything in properties secion, Obsidian will completely restructure the entire frontmatter content.

Codeblocks look safer. But I still would like to see them as yaml)

On the other hand, with frontmatter as the storage we could get automatic translation yaml to json, and I believe json to yaml too, because if you put something like

```markdown
---
{ "hello": "world" }
---
```

-Obsidian can parse that.

Also there can only be one frontmatter per note. Is that a pro or a con - I'm not sure. Maybe there are people who configure the Obsidian Web slightly differently on different hosts.

**Reply by coddingtonbear** [reply]
The reason Obsidian can parse that is that JSON is a subset of YAML -- so: all YAML parsers can understand JSON, including the one we use for frontmatter.

That being said -- I'm surprised you folks are talking about putting this in frontmatter or a code block; what's the value of doing that over just placing the content directly in a file in your vault?

**Reply by 0xdade** [reply]
I would actually be totally fine putting the .json config file directly in a file in my vault. My only minor complaint about that is that obsidian can't open `.json` files and doesn't show them in the sidebar by default. But also for my purposes, just automatically saving/loading the config file to a custom path would solve my wants.

**Reply by coddingtonbear** [reply]
I'd be sort of surprised if there were ever a reason to look at the config file -- I definitely do not structure it with the intent of it being read by a human, and have to say that the idea of folks hand-modifying one makes me a little itchy because of all of the difficult-to-troubleshoot bugs that kind of thing might inspire.
